lib/rouge/lexers/csharp.rb

```diff
@@ -81,14 +81,15 @@ class CSharp < RegexLexer
         rule %r/@"(""|[^"])*"/m, Str
         rule %r/"(\\.|.)*?["\n]/, Str
         rule %r/'(\\.|.)'/, Str::Char
-        rule %r/0x[0-9a-f]+[lu]?/i, Num
+        rule %r/0b[_01]+[lu]?/i, Num
+        rule %r/0x[_0-9a-f]+[lu]?/i, Num
         rule %r(
-          [0-9]
-          ([.][0-9]*)? # decimal
-          (e[+-][0-9]+)? # exponent
-          [fldu]? # type
+          [0-9](?:[_0-9]*[0-9])?
+          ([.][0-9](?:[_0-9]*[0-9])?)? # decimal
+          (e[+-]?[0-9](?:[_0-9]*[0-9])?)? # exponent
+          [fldum]? # type
```

Comment:
@jneen I'm curious on your thoughts here. I always recall you saying that a lexer doesn't necessarily have to be syntactically correct. In the case of this regex is there any reason we shouldn't accept the extended regex here for correctness? Or could we, for instance, collapse some of this and says an `_` at the beginning or end is OK, although syntactically incorrect for C#?

```ruby
          [_0-9]
          ([.][_0-9]*)? # decimal
          (e[+-][_0-9]+)? # exponent
          [fldum]? # type
```

Reply:
Totally understandable if you want to keep things simple. I would still suggest to keep the exponent sign optional (otherwise not all numbers will match the pattern)... and perhaps require numbers to always start with a digit (to be able to classify numbers vs. identifiers on the very first character, not sure if that matters a lot here though).